### PR TITLE
Augment-ed *Entry and Use'd *Entry stored in parent *Entry's new Augmented and Uses fields.

### PR DESCRIPTION
**Pull request by LeonGWang**
This perserves the Augment *Entry and Uses *Entry so their when statements can be enforced, among other things.

This is my attempt at fixing issue #89

### CODE REVIEW

File: pkg/yang/entry.go [hunk 1]
```diff
@@ -925,6 +925,11 @@ func (e *Entry) Augment(addErrors bool) (processed, skipped int) {
 		// are merged into another entry.
 		processed++
 		ae.merge(nil, a.Namespace(), a)
+		if ae.Annotation == nil {
+			ae.Annotation = map[string]interface{}{}
+			ae.Annotation["Augmentation"] = []*Entry{}
```
[the remark this answers is not in the document]
Comment (LeonGWang) [reply pkg/yang/entry.go]: I agree. I will make this a top-level member of the `Entry` type instead of being in the Annotation map

File: pkg/yang/entry.go [hunk 2]
```diff
@@ -925,6 +925,11 @@ func (e *Entry) Augment(addErrors bool) (processed, skipped int) {
...
+		}
+		ae.Annotation["Augmentation"] = append(ae.Annotation["Augmentation"].([]*Entry), a)
```
[the remark this answers is not in the document]
Comment (LeonGWang) [reply pkg/yang/entry.go]: For container and leaf `*Entry` we can get their when statements's XPath expression as follows:
`whenXPath, ok := *Entry.Node.(*Container/*Leaf).When.Statement().Arg()`
This is maybe a little verbose, but otherwise very workable.

 So I don't think a `type when struct` is necessary, because if we have the `*Entry`, then the `*Entry` == the `logicalProvenance` field, and the XPath `condition` can be extracted as demonstrated above. 

The problem with augment and uses is that the `*Entry` tree doesn't contain augment or uses `*Entry`, only the `*Entry` of the data nodes and/or groupings that are referred to by the augment and uses. Without access to augment and uses `*Entry` we also lose all information about them, including their when statements. So if we store augment and uses `*Entry` in the `*Entry` tree, then we have access to all the information the augment and uses `*Entry` have, not just the when statements.

What do you think?

**Comment by LeonGWang** [comment]
I have updated the Pull Request and the commit message to also include Uses *Entry.

File: pkg/yang/entry.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (robshakir) [reply pkg/yang/entry.go]: `*Entry.Node.(*Container).When.Statement().Arg()` is a bit wordy -- but we can probably help this by just having a helper that returns this on a particular `*Entry` receiver. So this sounds OK to me.

I have a concern with just copying the `*Entry` here though -- at this point, we've now included all the hierarchy below the `augment` `Entry` again in the tree. For example, if we have these modules:

```yang
module test {
  prefix "t";
  namespace "urn:t";

  container base { }
}
```

```yang
module test2 {
  prefix "t2";
  namespace "urn:t2";

  import test { prefix t; }
  
  augment /t:base {
    container a {
      container b {
        container c {
          leaf d { type string; }
        }
      }
    }
  }
}
```

Then the entry at `/t:base`'s `Augmented` will contain the `augment t:base` entry, and its `/t:base`'s `Dir` will also contain the same data. This will be for the entire tree.

It seems like it might be more efficient to, rather than store the whole entry, consider whether we can store a subset of it. (Note this information is something that we need to store in the JSON serialisation too - so there's a cost for tools that are using this to persist the processed tree - like ygot - so this is why I am concerned with the size.)

There are two options that spring to mind:

1) Store only the direct children of a particular Entry in the `Dir` of the entry which is included in `Augmented` or `Uses`.
2) Rather than store the `Entry` in `Augmented` -- add the `when` statements directly to entries that are included due to the `uses` or `augment`.

The latter approach would mean that:

```
container a {
   uses foo { when "something" }
}

grouping foo {
   leaf a { type string ; }
   container b { 
    leaf c { type string; }
   }
}
```

We would end up appending the `when` statement to the entry for `a` and `b` -- this results in duplication of the `when` contents -- and potentially processing by other tools.

In the former, we'd include only `foo` as an `Entry` in the `Uses` slice, but `c` would not be included. More efficient for processing - and avoids duplication of the entire tree when we serialise. We could actually remove anything in the `Entry` other than the children's name in this case.

[the remark this answers is not in the document]
Comment (LeonGWang) [reply pkg/yang/entry.go]: I have added a helper function `func (*Entry) GetWhenXPath() (string, bool)` as a shortcut for `*Entry.Node.(*TypeThatHaveWhenStatements).When.Statement().Arg()`

I went with option 1: Store only the direct children of a particular Entry in the `Dir` of the entry which is included in `Augmented` or `Uses`. I added a helper function `shallowDup` for this purpose.

Thanks for the timely and detailed responses. Let me know what you think of the latest version.

**Comment by LeonGWang** [comment]
I have updated the Pull Request to increase code coverage for the helper function `GetWhenXPath`

**Comment by robshakir** [comment]
@LeonGWang -- are you waiting on anything here from me? I think there were just minor comments to resolve.

**Comment by LeonGWang** [comment]
I am back from PTO, let me know if anything needs to be done one this commit.

File: pkg/yang/entry.go [hunk 3]
```diff
@@ -95,9 +95,11 @@ type Entry struct {
 	// is a module only.
 	Identities []*Identity `json:",omitempty"`
 
-	Augments   []*Entry                   `json:"-"` // Augments associated with this entry.
+	Augments   []*Entry                   `json:"-"` // Augments defined in this entry.
+	Augmented  []*Entry                   `json:"-"` // Augments merged into this entry.
 	Deviations []*DeviatedEntry           `json:"-"` // Deviations associated with this entry.
 	Deviate    map[deviationType][]*Entry `json:"-"`
+	Uses       map[*Uses]*Entry           `json:"-"` // Uses merged into this entry
```
[the remark this answers is not in the document]
Comment (LeonGWang) [reply pkg/yang/entry.go]: Looks like JSON only supports the `string` type as map keys. What do you think is a good solution to encode the `map[*Uses]*Entry` map? Should I encode `*Uses `as its field `Name`, which is type `string` and represents the name of the used `grouping`?

[the remark this answers is not in the document]
Comment (robshakir) [reply pkg/yang/entry.go]: One option here would be to change to make this something like:

```golang
type UsesStmt struct {
  Uses *Uses
  Grouping *Entry
}

type Entry struct {
  Uses []*UsesStmt
}
```

this would remove the need for the `map` here, to replace it with a slice. The downside of doing this is that we can't look up based on a `Uses` any more (however, this seems unlikely that one would do this, since it needs a pointer to the actual `Entry` that we don't store).

I think its important that we can persist the whole `Uses` since it gives us a bunch of additional information rather than just the grouping name.

**Comment by robshakir** [comment]
Apologies for the delay -- I've also been on PTO. I'd just like to figure out the map vs. slice for `Uses` and then we're good to merge this.

File: pkg/yang/entry.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (LeonGWang) [reply pkg/yang/entry.go]: That's a good idea. I will give it a try.

[the remark this answers is not in the document]
Comment (LeonGWang) [reply pkg/yang/entry.go]: Ok, I have converted Uses to []*UsesStmt and update marshal_test.go. Please let me know what you think!

**Comment by LeonGWang** [comment]
I made `Uses` into a slice so JSON marshalling will work. Please let me know what you think!

**Comment by LeonGWang** [comment]
The newest commit also fixes issue #91

**Comment by robshakir** [comment]
Great, thank you so much for all the work on this -- and the iteration after review comments. Let's merge this! :ship: :shipit:

**Comment by LeonGWang** [comment]
Thanks for all the help and pointers along the way Rob!
